types/datum.go

```diff
@@ -1478,6 +1482,14 @@ func (d *Datum) toSignedInteger(sc *stmtctx.StatementContext, tp byte) (int64, e
 		return ConvertJSONToInt(sc, d.GetMysqlJSON(), false)
 	case KindBinaryLiteral, KindMysqlBit:
 		val, err := d.GetBinaryLiteral().ToInt(sc)
+		if sc.CastInInsert {
```

Comment:
It seems this PR(https://github.com/pingcap/tidb/pull/11106) fixes the same problem... and it doesn't take statement context into account. @b41sh  PTAL.

Reply:
After discussing in #11106, is #11106 more appropriate than this PR?
What can we do for this PR, and if it's nothing to do, can we close this PR?

Reply:
After discussing in #11106, is #11106 more appropriate than this PR?
What can we do for this PR, and if it's nothing to do, can we close this PR?

Reply:
> After discussing in #11106, is #11106 more appropriate than this PR?
> What can we do for this PR, and if it's nothing to do, can we close this PR?

If #11106 is more appropriate than this PR, it should pass the all tests in make test. I am curious that if there are other ways to pass the all tests.
Note that it fail to deal with `select cast(0xffffffffffffffffffffffffff as signed);` which should return `-1`;